plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/findReferences/XWorkspaceResourceAccess.java

```diff
@@ -25,27 +25,37 @@
 @SuppressWarnings("restriction")
 public class XWorkspaceResourceAccess implements IReferenceFinder.IResourceAccess {
 
-	private final XWorkspaceManager workspaceManager;
+	private final XLanguageServerImpl languageServer;
 
 	/**
-	 * @param workspaceManager
-	 *            the workspace manager
+	 * @param languageServer
+	 *            the language server
 	 */
-	public XWorkspaceResourceAccess(XWorkspaceManager workspaceManager) {
-		this.workspaceManager = workspaceManager;
+	public XWorkspaceResourceAccess(XLanguageServerImpl languageServer) {
+		this.languageServer = languageServer;
 	}
 
 	@Override
 	public <R> R readOnly(URI targetURI, IUnitOfWork<R, ResourceSet> work) {
-		XProjectManager projectManager = workspaceManager.getProjectManager(targetURI);
-		if (projectManager == null) {
+		OpenFilesManager openFilesManager = languageServer.getOpenFilesManager();
+		if (openFilesManager == null) {
 			return null;
 		}
+		// FIXME GH-1774 consider re-using the resource set of the current open file context for other files:
+		// OpenFileContext currOFC = openFilesManager.currentContext();
+		// if (currOFC != null) {
+		// return doWork(currOFC.getResourceSet(), work);
+		// }
+		return openFilesManager.runInOpenOrTemporaryFileContextSync(targetURI, "XWorkspaceResourceAccess",
```

Comment:
Two readonly may be nested in each other with different targetURIs. If the are from different projects, by transitivity we may have loaded two resources with the same URI into two different resource sets. I'm skeptical that we won't run into identity issues.

Reply:
The idea is that the `#runInOpen|TemporaryFileContext()` methods are used only in two cases:
1. by the LSP framework (mostly XLanguageServerImpl, maybe also N4JSCodeActionService / N4JSCommandService) to kick off the computation of some editing functionality,
2. somewhere nested inside editing functionality to run a computation in a resource entirely separate from the "main resource" of the open editor the editing functionality originated in.

No. 2 is expected to be a rather rare case and used only for operations that are self-contained and do not exchange `EObject`s (esp. `TModule` elements) with the outside. An example might be looking up the JSDoc string of a method in some arbitrary N4JS[D] file.

The standard case of using information of a resource B from within the context of an open file A is to simply load resource B in the resource set of the open file A (as usual with EMF and as the Eclipse editors are doing it usually).

So far my plan/intention.

HOWEVER: the current invocations of `#runInOpen|TemporaryFileContext()` on my branch are partly incorrect and thus give a wrong impression of the intention of those methods. The reason is that I haven't fully implemented loading of resources from source in the resource sets of open files (no `IExternalContentSupport` registered on the resource sets in `OpenFileContext` yet) and I have just added the `#runInOpen|TemporaryFileContext()` as "place holders", because I have to revisit those anyway.

Reply:
> no IExternalContentSupport registered

Understood.